salt/modules/netbox.py

```diff
@@ -117,10 +187,980 @@ def get(app, endpoint, id=None, **kwargs):
+    payload = {'model': model, 'manufacturer': nb_man['id'], 'slug': slugify(model)}
+    typ = _add('dcim', 'device-types', payload)
+    ret_dict = {'dcim': {'device-types': payload}}
+    if new_man:
+        ret_dict['dcim'].update(new_man['dcim'])
+    if typ:
+        return ret_dict
+    else:
+        return False
+
+
+def create_device_role(role, color):
+    '''
+    .. versionadded:: Fluorine
+
+    Create a device role
+
+    role
+        String of device role, e.g., ``router``
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt myminion netbox.create_device_role router
+    '''
+    nb_role = get_('dcim', 'device-roles', name=role)
+    if nb_role:
+        return False
+    else:
+        payload = {'name': role, 'slug': slugify(role), 'color': color}
+        role = _add('dcim', 'device-roles', payload)
+        if role:
+            return{'dcim': {'device-roles': payload}}
+        else:
+            return False
+
+
+def create_platform(platform):
+    '''
+    .. versionadded:: Fluorine
+
+    Create a new device platform
+
+    platform
+        String of device platform, e.g., ``junos``
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt myminion netbox.create_platform junos
+    '''
+    nb_platform = get_('dcim', 'platforms', slug=slugify(platform))
+    if nb_platform:
+        return False
+    else:
+        payload = {'name': platform, 'slug': slugify(platform)}
+        plat = _add('dcim', 'platforms', payload)
+        if plat:
+            return {'dcim': {'platforms': payload}}
+        else:
+            return False
+
+
+def create_site(site):
+    '''
+    .. versionadded:: Fluorine
+
+    Create a new device site
+
+    site
+        String of device site, e.g., ``BRU``
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt myminion netbox.create_site BRU
+    '''
+    nb_site = get_('dcim', 'sites', name=site)
+    if nb_site:
+        return False
+    else:
+        payload = {'name': site, 'slug': slugify(site)}
+        site = _add('dcim', 'sites', payload)
+        if site:
+            return {'dcim': {'sites': payload}}
+        else:
+            return False
+
+
+def create_device(name,
+                  role,
+                  model,
+                  manufacturer,
+                  site):
+    '''
+    .. versionadded:: Fluorine
+
+    Create a new device with a name, role, model, manufacturer and site.
+    All these components need to be already in Netbox.
+
+    name
+        The name of the device, e.g., ``edge_router``
+    role
+        String of device role, e.g., ``router``
+    model
+        String of device model, e.g., ``MX480``
+    manufacturer
+        String of device manufacturer, e.g., ``Juniper``
+    site
+        String of device site, e.g., ``BRU``
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt myminion netbox.create_device edge_router router MX480 Juniper BRU
+    '''
+    try:
```

Comment:
This is duplicated in a few places. Can we refactor it into a single location?

Reply:
From what I understood, you want to refactor the "does this object exist in netbox" checks to a single location. Most of these checks are single use, except for the device one, which I guess could be put in a function. Is this what you meant?